### PR TITLE
feat: Add "slash validator" gov proposal

### LINKED ISSUE #64: Make a slash governance proposal
## Summary

There should be a governance proposal type to slash a validator. It would specify `slash_percentage` and `validator address`.

## Problem Definition

This has a number of use cases. It allows blockchain communities to codify social standards that validators have to adhere to, and in the event they are 'caught' by the community, it allows them to slash them on-chain. (Without needing a hard fork)

A social slashing standard that has interest from the Osmosis community is essentially a "No Sybilling" condition, where they can all agree that if governance is presented with sufficiently persuasive evidence of a validator sybilling in order to game airdrops, newer staking reward designs, etc. they would be slashed for a certain factor. (With some carveout for cotenancy situations etc.). This is in part motivated by some validator-sybil-gameable decentralization ideas, cref https://commonwealth.im/osmosis/discussion/2036-proposal-per-validator-epoch-bonus-for-better-decentralization

I'm generally of the view that this is important for many chains, if validators commit malicious activities not yet thought about in advance, or on other chains. I propose we add this as a governance proposal type to the slashing module. Chains can simply choose not to add this governance proposal type to their governance router if they don't want this feature.

## Proposal

Make a new governance proposal type in slashing with the following proto:
```
message SlashValidatorProposal {
  option (gogoproto.equal) = true;
  option (gogoproto.goproto_getters) = false;
  option (gogoproto.goproto_stringer) = false;

  string title = 1;
  string description = 2;
  string  validator_address = 3 [(cosmos_proto.scalar) = "cosmos.AddressString"];
  string slash_factor_percent = 4 [
    (cosmos_proto.scalar)  = "cosmos.Dec",
    (gogoproto.customtype) = "github.com/cosmos/cosmos-sdk/types.Dec",
    (gogoproto.nullable)   = false
  ];
}
```

Heres an example of how you do custom governance proposal types if folks are interested: 
* proto: https://github.com/osmosis-labs/osmosis/blob/main/proto/osmosis/pool-incentives/v1beta1/gov.proto
* gov.go: https://github.com/osmosis-labs/osmosis/blob/main/x/pool-incentives/types/gov.go
* keeper: https://github.com/osmosis-labs/osmosis/blob/main/x/pool-incentives/keeper/gov.go
* app.go: https://github.com/osmosis-labs/osmosis/blob/main/app/keepers.go#L263

Happy to help guide anyone interested in implementing this

To do this, you'd implement:
* the proto file here: https://github.com/cosmos/cosmos-sdk/tree/master/proto/cosmos/slashing/v1beta1
* gov.go here: https://github.com/cosmos/cosmos-sdk/tree/master/x/slashing/keeper
  * basically just calling https://github.com/cosmos/cosmos-sdk/blob/2f335a2b8cc4fcfa337c104823fab9f6ecb387da/x/slashing/keeper/keeper.go#L68 with distributionheight = ctx.BlockHeight()
* Another gov.go in the types folder

## Reference SDK issue

There is a reference issue in the SDK here https://github.com/cosmos/cosmos-sdk/issues/10917 . I expect that I'm only going to track this one for the foreseeable future, since that one is blocked on a multi-month gov refactor.

____

#### For Admin Use

- [ ] Not duplicate issue
- [ ] Appropriate labels applied
- [ ] Appropriate contributors tagged
- [ ] Contributor assigned/self-assigned

### PR DESCRIPTION
**Pull request by hleb-albau**
## Description

Closes: #64

Adds "slash validator" gov proposal

---

### Author Checklist

*All items are required. Please add a note to the item if the item is not applicable and
please add links to any relevant follow up issues.*

I have...

- [ ] included the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ] added `!` to the type prefix if API or client breaking change
- [ ] targeted the correct branch (see [PR Targeting](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#pr-targeting))
- [ ] provided a link to the relevant issue or specification
- [ ] followed the guidelines for [building modules](https://github.com/cosmos/cosmos-sdk/blob/master/docs/building-modules)
- [ ] included the necessary unit and integration [tests](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#testing)
- [ ] added a changelog entry to `CHANGELOG.md`
- [ ] included comments for [documenting Go code](https://blog.golang.org/godoc)
- [ ] updated the relevant documentation or specification
- [ ] reviewed "Files changed" and left comments if necessary
- [ ] confirmed all CI checks have passed

### Reviewers Checklist

*All items are required. Please add a note if the item is not applicable and please add
your handle next to the items reviewed if you only reviewed selected items.*

I have...

- [ ] confirmed the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ] confirmed `!` in the type prefix if API or client breaking change
- [ ] confirmed all author checklist items have been addressed 
- [ ] reviewed state machine logic
- [ ] reviewed API design and naming
- [ ] reviewed documentation is accurate
- [ ] reviewed tests and test coverage
- [ ] manually tested (if applicable)

### CODE REVIEW

**Comment by hleb-albau** [comment]
@ValarDragon seems ready, all check that failing, do it in base branch too.

**Comment by ValarDragon** [comment]
TYSM!

yeah that test-race failing is fine (its not an actual race condition) 

Will review ASAP!

File: docs/core/proto-docs.md [hunk 1]
```diff
@@ -5884,6 +5888,24 @@ Params represents the parameters used for by the slashing module.
 
 
 
+<a name="cosmos.slashing.v1beta1.SlashValidatorProposal"></a>
+
+### SlashValidatorProposal
+
+
```
Comment (ValarDragon) [inline docs/core/proto-docs.md:5895]: Could you briefly describe the edge cases here? (E.g. can also slash inactive proposers, slash happens to live stakers + unbonders at time of vote end)

File: x/slashing/proposal_handler.go [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+		return types.ErrSlashUnbondedValidator
+	}
+
+	valConsAddr, valConsdAddrErr := validator.GetConsAddr()
+	if valConsdAddrErr != nil {
+		return valConsdAddrErr
+	}
+
+	valPower := stakingKeeper.GetLastValidatorPower(ctx, valAddr)
+	slashingKeeper.Slash(ctx, valConsAddr, p.SlashFactor, valPower, ctx.BlockHeight())
```
Comment (ValarDragon) [inline x/slashing/proposal_handler.go:49]: Do you think we should emit an event here as well? (TBH its weird that the slash function doesn't have a "reason" field)

File: docs/core/proto-docs.md (hunk 1 above)
Comment (hleb-albau) [reply docs/core/proto-docs.md:5895]: okey, make sense

File: x/slashing/proposal_handler.go (hunk 2 above)
Comment (hleb-albau) [reply x/slashing/proposal_handler.go:49]: agree, nice catch!

File: x/slashing/proposal_handler_test.go [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+	"github.com/cosmos/cosmos-sdk/x/slashing/types"
+	"github.com/cosmos/cosmos-sdk/x/staking"
+	"github.com/cosmos/cosmos-sdk/x/staking/teststaking"
+	"github.com/stretchr/testify/require"
+	tmproto "github.com/tendermint/tendermint/proto/tendermint/types"
+	"testing"
+)
+
+func TestSlashValidatorProposal(t *testing.T) {
+	// initial setup
```
Comment (ValarDragon) [inline x/slashing/proposal_handler_test.go:16]: Can we add another comment for whats the setup logic / what its testing?

Appears to me: Setup a staking module with max 1 validator,  and then two validators in queue. Run a block. Slash the validator thats not in the active set via proposal (it fails)

Slash the validator in the active set (it succeeds)

What I don't get / should we be testing: Does the #2 validator then become the #1 validator and the first validator is jailed, or is this not jailing?

File: x/slashing/proposal_handler_test.go [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+	require.NoError(t, err)
+
+	updates, err := app.StakingKeeper.ApplyAndReturnValidatorSetUpdates(ctx)
+	require.NoError(t, err)
+	require.Equal(t, 1, len(updates))
+
+	slashedValidator, _ := app.StakingKeeper.GetValidatorByConsAddr(ctx, sdk.GetConsAddress(pks[0]))
+	if slashedValidator.Tokens.GTE(activeValidator.Tokens) {
+		t.Fail()
+	}
```
Comment (ValarDragon) [inline x/slashing/proposal_handler_test.go:62]: can we check using the exact slash factor?

**Comment by ValarDragon** [comment]
This looks great to me, thanks so much for the swift PR! I think it'd be nice if we could improve the test a bit (some more docs / ideally making it more table driven if you have bandwidth, since theres a bunch more edge cases to test for). 

We need to test that unbonding validators can be slashed. 

Also this is a point you raised in the original issue - what height do we slash at. Originally I think I said the current height is fine. 

After thinking about it more, it definitely needs to be a height at proposal creation (or ideally even before proposal creation, since it'd likely be talked about on a governance platform for a couple days). Otherwise everyone would start unbonding immediately or redelegate away, and not get slashed.

Can we make the governance proposal include a slashing block height? (And we explain in docs that this needs to be set in the block range `[proposal_getting_on_chain_height - unbonding_period_in_blocks + voting_period_in_blocks, proposal_getting_on_chain_height]` )

One other question I have is, do you think we should make the slash proposal be able to specify multiple validators? Something I didn't really think about in the initial proposal but I think it'd be a useful feature if the goal is to be able to slash for sybilling. (Want to slash both identities at once)

File: x/slashing/proposal_handler_test.go (hunk 3 above)
Comment (hleb-albau) [reply x/slashing/proposal_handler_test.go:16]: yeah, in general i spent a lot of time reading slashing tests and trying understand them myself.
Slashing module itself is nightmare from my point of view, lots of panic and other staff.
Here is copy-paste from test slash module. In general i don't understand why there are not test setups helpers that can be used in other modules. But osmo fork is not that place, where it should be fixed :(

Will add more comments here about our test, to make it clear whats going on.

It is not jailing, only slashing. Should we consider jailing as proposal option?

File: x/slashing/proposal_handler_test.go (hunk 4 above)
Comment (hleb-albau) [reply x/slashing/proposal_handler_test.go:62]: let's try, but for some reason slash tests use such checks(maybe there is cases when slash can be lower)

**Comment by hleb-albau** [comment]
> 

I will recheck, but seems if vote time is lower than unbonding time, than all undelegations and redelegations will be slashed.

Yeah, multiple validators make sense.

In general, issue seems a bit more complicated and deeper, than i expected. Need more caution  here :)

**Comment by ValarDragon** [comment]
Definitely agreed, it was more complex than I expected as well

**Comment by ValarDragon** [comment]
So it has to slash using a historical block height in order to slash the undelegations and redelegations. (Which is why I was suggesting providing the block height in the proposal itself, since I think there may be a lot factors that are relevant in choosing it)

At the moment, on this line: https://github.com/osmosis-labs/cosmos-sdk/pull/85/files#diff-fe15689ebfe9ff4870f298cae385f07ab644f0a0938652fb465bde220de3532cR49 , its slashing at the current height. (ctx.BlockHeight())

If we look into the slash function, it actually acts differently for slashes at the current height vs at historical heights. https://github.com/osmosis-labs/cosmos-sdk/blob/v0.45.0x-osmo-v7/x/staking/keeper/slash.go#L67-L81

If your slashing at the current height, all people undelegating and redelegating left "before" the bad event which your getting slashed for happened, so they don't get slashed.

File: x/slashing/proposal_handler_test.go (hunk 3 above)
Comment (faddat) [reply x/slashing/proposal_handler_test.go:16]: probably, there could be circumstances where a validator needs to be removed from the active set.

Comment (ValarDragon) [reply x/slashing/proposal_handler_test.go:16]: Oh yikes, didn't realize the slashing tests weren't well architected to begin with :/

**Comment by hleb-albau** [comment]
While developing new version of proposal I found that it is not easy to add slash height on proposal.

1. The main problem here is we need to store validator power(or tokens) for given height. As I saw in double sign evidence, they provide it in message (btw there is now real validation of historical power value, thus i think it is potential exploit to over-slash validator).
So i think we should not add  index to staking module in context of this fork, but rather keep it simple - slash at current height, but add one more param to slash function - slash delegation at current height.
2. For unbonding case slash function slashes only `NotBondedTokens`. Thus we should modify slash function, or consider something else.
3. Should we set `jailUntil` time, or allow to unjail immediately (like in case of missed blocks)

Also, writing tests here is more time consuming, rather than feature itself :( Seems we should test only our part here, relying on slash logic as valid.

I start to worry about changing sdk too much - thus hardening next merges with upstream. Maybe cut some logic here(like unbounding or delegations), cuz for minor purposes even jailing will be enough. 

waiting your feedback

**Comment by faddat** [comment]
@hleb-albau hi!

I missed this pr into the osmosis-sdk. 

I've been toying with it tonight, but suppose I won't continue right now. 

"slash at current height" so that'd be a slash, at the time that the proposal passes.  I am trying to think like a bandit.  

* I think that this should set JailUntil time, possibly as a parameter
* if I understand it right, `NotBondedTokens` shifts instantly at the time of unbonding.  If this is to be used as a tool to drive people away from a validator, I think that this is sufficient.  Probably better than rewriting its dependencies. 

https://github.com/cosmos/cosmos-sdk/blob/58597139fa0fb9e9be60deebee3df1663aa2cfaf/x/staking/spec/02_state_transitions.md?plain=1#L34-L43

**opinion**:

I'm considering another airdrop for dig, and I think that the way I am thinking about it, might help to inform the deisgn of this gov proposal.  

We're really focused on governance and so it'll go to:

* users who participated in gov on osmo and juno
* users who delegated to validators who participated in gov on osmo and juno

Basically I'm using the "representative" concept of the validator, and indeed, I would not give tokens to those who chose validators who do not vote.

So I guess I am saying that with respect to users, if it is possible to slash users in a manner that's equal 1:1 %age wise to how the validator is slashed, I think that is going to have the best possible effect on user choices of validators. 

**thanks for your work**

~~I guess that one of your concerns is that users can redelegate rapidly, and thus avoid the slash?~~

https://github.com/cosmos/cosmos-sdk/blob/58597139fa0fb9e9be60deebee3df1663aa2cfaf/x/staking/spec/02_state_transitions.md?plain=1#L116

Guess not?

**Comment by p0mvn** [comment]
Hi everyone! What is the status of this? I see that the linked issue was closed. Can this PR be closed as well or is the work ongoing?

**Comment by p0mvn** [comment]
Please convert this to a draft if it is still in progress. Otherwise, let's close

**Comment by hleb-albau** [comment]
lets close it for now
seems it will be implemented in sdk itself
